### PR TITLE
Module loggers w/ individually configurable levels

### PR DESCRIPTION
**Pull request by jwolski**
Some apps like to log. Some apps don't. Some logs are nice to have for debugging. Some logs aren't. Module loggers allow different Ringpop components/modules to have individually configurable logging levels which are settable in the Ringpop constructor or over the /admin/config/set endpoints.

reviewers: @thanodnl @benfleis 
cc: @uber/ringpop

### CODE REVIEW

File: lib/logging/module_logger.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var Levels = require('./level.js');
```
Comment (thanodnl) [inline lib/logging/module_logger.js]: levels.js

Comment (jwolski) [inline lib/logging/module_logger.js]: thx

File: lib/gossip/ping-req-sender.js [hunk 2]
```diff
@@ -184,7 +184,7 @@ module.exports = function sendPingReq(opts, callback) {
     for (var i = 0; i < pingReqMembers.length; i++) {
         var pingReqMember = pingReqMembers[i];
 
-        gossipLogger(ringpop, 'ringpop ping-req send', {
+        this.gossipLogger.info('ringpop ping-req send', {
```
Comment (thanodnl) [inline lib/gossip/ping-req-sender.js]: This is outside of the scope of the `PingReqSender`

Comment (jwolski) [inline lib/gossip/ping-req-sender.js]: thx. u r correct. i will fix.

File: config.js [hunk 3]
```diff
@@ -65,9 +66,13 @@ Config.prototype._seed = function _seed(seed) {
     // All config names should be camel-cased.
     seedOrDefault('TEST_KEY', 100, numValidator); // never remove, tests and lives depend on it
 
+    // Logger configs
+    seedOrDefault('defaultLogLevel', LoggingLevels.info);
+    seedOrDefault('gossipLogLevel', LoggingLevels.off);
+    seedOrDefault('joinerLogLevel', LoggingLevels.warn);
```
Comment (benfleis) [inline config.js]: reading 'joiner' makes me wonder -- is level setting intended to apply to objects (Joiner) or concepts (join)?

if the latter, then perhaps it should be s/joiner/join/ to match other join knobs?

File: lib/logging/levels.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+var levels = ['trace', 'debug', 'info', 'warn', 'error', 'off'];
+
+// e.g. trace => 'trace', debug => 'debug', etc.
+var namedLevels = levels.reduce(function each(res, level) {
+    res[level] = level;
+    return res;
+}, {});
+
+// e.g. trace => 0, debug => 1, etc.
+var levelToInt = {};
```
Comment (benfleis) [inline lib/logging/levels.js]: Nit: IMHO there is value in corresponding to [syslog standards](https://en.wikipedia.org/wiki/Syslog#Severity_level) for numbering. Maybe it never matters due to export, but my €0,02.

File: lib/logging/logger_factory.js [hunk 5]
```diff
@@ -19,15 +19,33 @@
...
+        case 'gossip':
+            this.loggers[name] = new ModuleLogger(this.ringpop,
+                'gossipLogLevel');
+            break;
+        case 'joiner':
+            this.loggers[name] = new ModuleLogger(this.ringpop,
+                'joinerLogLevel');
+            break;
+        default:
+            this.loggers[name] = new ModuleLogger(this.ringpop,
```
Comment (benfleis) [inline lib/logging/logger_factory.js]: this is boss for adding temporary/local logging, and still being able to control it :)

Comment (benfleis) [inline lib/logging/logger_factory.js]: on second read (and reading further down), could this be better handled as a fallthrough, so that I can set foobarLogLevel in my config, and have it still work?

**Comment by benfleis** [comment]
added a couple thoughts, trlgtm.

File: config.js (hunk 3 above)
Comment (jwolski) [inline config.js]: fair. 'packages' or 'modules' is a good delineation. will fix.

File: lib/logging/levels.js (hunk 4 above)
Comment (jwolski) [inline lib/logging/levels.js]: interesting. i approve.

File: lib/logging/logger_factory.js (hunk 5 above)
Comment (jwolski) [inline lib/logging/logger_factory.js]: lemme think about this.

File: lib/logging/levels.js (hunk 4 above)
Comment (jwolski) [inline lib/logging/levels.js]: on second thought, while it's nice to reflect some established standards, these int values are completely hidden by Ringpop. There are log levels that dont really apply to the logger Ringpop depends on e.g. critical, notice, emergency. I'm going to leave as is.

File: lib/logging/logger_factory.js (hunk 5 above)
Comment (jwolski) [inline lib/logging/logger_factory.js]: i thought about it and will keep as is until we require such flexibility.
